### PR TITLE
refactor: FPE safe eta/theta conversion

### PR DESCRIPTION
**Pull request by andiwand**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new structure for angle conversion traits, enhancing precision and safety for angle calculations.
	- Added safety checks in angle conversion functions to handle edge cases more gracefully.

- **Bug Fixes**
	- Updated angle conversion functions to prevent floating-point exceptions by implementing bounds checking.

- **Tests**
	- Added new data-driven test cases to validate the robustness of angle conversion functions, ensuring consistent and valid outputs across a range of inputs.

### CODE REVIEW

**Comment by andiwand** [comment]
See two open discussion items here
- https://github.com/acts-project/acts/pull/3756#discussion_r1817429988
- https://github.com/acts-project/acts/pull/3756#discussion_r1817437660

**Comment by andiwand** [comment]
I experimented with the unit test and thrown FPEs to come up with some bounds

**Comment by paulgessinger** [comment]
Empirical boundaries seem fine to me personally.

**Comment by timadye** [comment]
> Empirical boundaries seem fine to me personally.

I don't know if this is what @paulgessinger meant.

Personally, I would prefer consistent boundaries, as I suggested in https://github.com/acts-project/acts/pull/3756#discussion_r1817850323 so you get

1. roughly the same limits for different FP types, and
2. if beyond the limit, you are returned the equivalent value at the limit, not &pm;`inf`.

I am less sure about (2). I don't know if any other bits of code (eg. printout of &eta;) worry about `inf`.

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 1]
```diff
@@ -9,26 +9,69 @@
...
 namespace Acts::AngleHelpers {
 
+template <std::floating_point Scalar>
+struct EtaThetaConversionTraits {
+  static constexpr Scalar minTheta = 0;
+  static constexpr Scalar maxTheta = std::numbers::pi_v<Scalar>;
+
+  static constexpr Scalar minEta = -std::numeric_limits<Scalar>::infinity();
+  static constexpr Scalar maxEta = std::numeric_limits<Scalar>::infinity();
+};
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:24]: I wonder if this may be a hostage to fortune if we have other `std::floating_point` types in future (eg. `long double`, `std::float16_t`, GPU types). Those conversions could still return `nan` rather than `inf`. I think any use of this template that is not specialized could be problematic.

Comment (andiwand) [reply Core/include/Acts/Utilities/AngleHelpers.hpp:24]: https://github.com/acts-project/acts/blob/6797a56a15e67cc036c15cfc623558d2e0286990/Core/include/Acts/Utilities/AlgebraHelpers.hpp#L217-L250

I guess we can do it like here and force a compile error in this case

**Comment by andiwand** [comment]
> > Empirical boundaries seem fine to me personally.
> 
> I don't know if this is what @paulgessinger meant.
> 
> Personally, I would prefer consistent boundaries, as I suggested in [#3756 (comment)](https://github.com/acts-project/acts/pull/3756#discussion_r1817850323) so you get
> 
> 1. roughly the same limits for different FP types, and
> 2. if beyond the limit, you are returned the equivalent value at the limit, not ±`inf`.
> 
> I am less sure about (2). I don't know if any other bits of code (eg. printout of η) worry about `inf`.

I don't see a good reason for making the bounds the same TBH. If `double` has more precision than `float` why not use it. Same goes for different bounds in different directions. If `eta->theta` has different numerical properties than `theta->eta` why trim the precision?

**Comment by paulgessinger** [comment]
What I meant is that pragmatically, for a selector like this, is there a reason that makes picking a very large value but not necessarily the numerical limit problematic. 

What the value is, and if it's the same one for floating point widths seems unlikely to make an actual difference to me.

**Comment by andiwand** [comment]
I added a clamped version @paulgessinger @timadye 

@timadye can you check if this is what you had in mind?

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 2]
```diff
@@ -9,26 +9,120 @@
...
+  static constexpr double minTheta = 1e-12;
+  static constexpr double maxTheta = std::numbers::pi - 1e-12;
+
+  static constexpr double minEta = -700.0;
+  static constexpr double maxEta = 700.0;
+};
+
+template <typename Scalar>
+struct ClampedEtaThetaConversionTraits {
+  // computed from minEta
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:39]:
```suggestion
  // computed from maxEta
```

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 3]
```diff
@@ -9,26 +9,120 @@
...
+  static constexpr Scalar minTheta =
+      static_cast<Scalar>(3.6097027756908306e-35);
+  // computed from maxEta
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:42]:
```suggestion
  // computed from minEta
```

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 4]
```diff
@@ -9,26 +9,120 @@
...
+
+  static constexpr double minEta = -700.0;
+  static constexpr double maxEta = 700.0;
+};
+
+template <typename Scalar>
+struct ClampedEtaThetaConversionTraits {
+  // computed from minEta
+  static constexpr Scalar minTheta =
+      static_cast<Scalar>(3.6097027756908306e-35);
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:41]:
```suggestion
      static_cast<Scalar>(3.6097027756908303447e-35L);
```
might as use `long double` in case we ever use that for `Scalar`.

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 5]
```diff
@@ -9,26 +9,120 @@
...
+  // computed from maxEta
+  static constexpr Scalar maxTheta = static_cast<Scalar>(3.141592653589793);
+
+  static constexpr Scalar minEta = static_cast<Scalar>(-80.0);
+  static constexpr Scalar maxEta = static_cast<Scalar>(80.0);
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:46]: A `float` `maxTheta=3.141592653589793` gives `nan` when converted to &eta;. I think it would be better to clamp at &eta;=10.

```suggestion
  static constexpr Scalar minEta = static_cast<Scalar>(-10.0);
  static constexpr Scalar maxEta = static_cast<Scalar>(10.0);
```
and update `minTheta` and `maxTheta` above:
```
  static constexpr Scalar minTheta = static_cast<Scalar>(9.0799859462585550026e-05L);
  static constexpr Scalar maxTheta = static_cast<Scalar>(3.1415018537303306529L);
```

**Review by timadye — COMMENTED** [review]
yes, with some tweaks.

File: Core/include/Acts/Utilities/AngleHelpers.hpp (hunk 5 above)
Comment (andiwand) [reply Core/include/Acts/Utilities/AngleHelpers.hpp:46]: I cannot reproduce the `nan`. Isn't the value you are referring to checked against in `etaFromThetaClamed` so it has to produce `minEta`?

Comment (timadye) [reply Core/include/Acts/Utilities/AngleHelpers.hpp:46]: I checked with the non-clamped formula. The thing is, I don't know if the next-smallest value gives `nan`. I thought 10 would be safer, but if you are sure 80 is OK for `float`, then that's fine.

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 6]
```diff
@@ -9,26 +9,120 @@
...
+  static constexpr double maxEta = 700.0;
+};
+
+template <typename Scalar>
+struct ClampedEtaThetaConversionTraits {
+  // computed from maxEta
+  static constexpr Scalar minTheta =
+      static_cast<Scalar>(3.6097027756908303447e-35L);
+  // computed from minEta
+  static constexpr Scalar maxTheta = static_cast<Scalar>(3.141592653589793);
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:43]:
```suggestion
  static constexpr Scalar maxTheta = static_cast<Scalar>(3.1415926535897932385L);
```
sorry, I should have mentioned this one explicitly. This is for eta=-80.

File: Core/include/Acts/Utilities/AngleHelpers.hpp (hunk 5 above)
Comment (andiwand) [reply Core/include/Acts/Utilities/AngleHelpers.hpp]: opted for 10 now as the theta scale looks more appropriate for healthy floats

**Comment by andiwand** [comment]
@timadye can you approve if you are OK with the changes?

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 7]
```diff
@@ -9,26 +9,68 @@
...
 
 namespace Acts::AngleHelpers {
 
+template <typename Scalar>
+struct EtaThetaConversionTraits {};
+
+template <>
+struct EtaThetaConversionTraits<float> {
+  static constexpr float minTheta = 1e-6f;
+  static constexpr float maxTheta = std::numbers::pi_v<float> - 1e-6f;
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:22]: It's not a big deal, but this might be a bit better:
```suggestion
  static constexpr float maxTheta = std::numbers::pi_v<float> - minTheta;
```
to be sure of typos.

Similarly for the other specialisation.

File: Core/include/Acts/Utilities/AngleHelpers.hpp [hunk 8]
```diff
@@ -9,26 +9,68 @@
...
+
+  static constexpr float minEta = -80.0f;
+  static constexpr float maxEta = 80.0f;
```
Comment (timadye) [inline Core/include/Acts/Utilities/AngleHelpers.hpp:25]:
```suggestion
  static constexpr float maxEta = 80.0f;
  static constexpr float minEta = -maxEta;
```

**Review by timadye — COMMENTED** [review]
Some arguably clearer definitions. If you don't like them, let me know and I'll approve without further checks.

File: Core/include/Acts/Utilities/AngleHelpers.hpp (hunk 7 above)
Comment (andiwand) [reply Core/include/Acts/Utilities/AngleHelpers.hpp:22]: sure - my thought was that the min/max would not have to match but as it is for now they do so we can just deduplicate the numbers

**Review by timadye — APPROVED** [review]
Looks good. My last comment was also made by CodeRabbit, so I guess I could be replaced by an AI 🤖. Actually I liked it's suggestion to make the conversions `constexpr`, so you could do that if you like. But let's finish with the comments now.

**Comment by andiwand** [comment]
I feat `constexpr` is only available with C++26 onwards for math functions.

**Comment by timadye** [comment]
> I feat `constexpr` is only available with C++26 onwards for math functions.

ah yes of course. I forgot I already mentioned that 😉

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [76bc018dfc5ecdc76d9f34cb38b4750dbb55fb16]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/8471769)!

|status|job|report|
|------|---|------|
|:red_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45717687)||
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716822)||
|:green_circle: | [run_art_test: test_run4_acts_vertex_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716821)||
|:green_circle: | [run_art_test: test_run4_ttbar_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716820)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716819)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716818)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716817)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716816)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716815)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716814)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716813)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716812)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716811)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716810)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716809)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716808)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716807)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716806)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716805)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716804)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716803)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716802)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716801)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716800)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716799)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716798)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716797)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716796)||
|:red_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716795)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716794)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716793)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716792)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716791)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716790)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/45716789)||
